tools/codegen/gen_functionalization_type.py

```diff
@@ -0,0 +1,344 @@
+    if len(f.func.arguments.out) != 0:
+        if len(f.func.arguments.out) > 1:
+            return_names = ', '.join(a.name for a in f.func.arguments.out)
+            return f'{DispatcherSignature.from_schema(f.func).returns_type().cpp_type()}({return_names});'
+        else:
+            return f.func.arguments.out[0].name
+    if f.func.arguments.self_arg is not None:
+        return f.func.arguments.self_arg.argument.name
+    raise AssertionError("Unable to handle functionalization for op={str(f.func.name)}")
+
+def is_multi_output(rets: Tuple[Return, ...]) -> bool:
+    return len(rets) > 1 or rets[0].type.is_list_like() is not None
+
+def arg_name(a: Argument) -> str:
+    return f'{a.name}_' if a.type.is_tensor_like() else a.name
+
+def unwrap_tensor_args_str(f: NativeFunction) -> str:
+    return '\n      '.join(
+        f'auto {arg_name(arg)} = at::functionalization::impl::from_functional_tensor({arg.name});'
+        for arg in f.func.arguments.flat_all if arg.type.is_tensor_like())
+
+# Generates the Functionalization kernel for:
+# - ops that create aliases (e.g. transpose())
+# - ops that are views AND mutations (e.g. transpose_())
+def emit_view_functionalization_body(f: NativeFunction, g: Optional[NativeFunctionsGroup], native_api_name: str) -> str:
+    dispatcher_sig = DispatcherSignature.from_schema(f.func)
+    dispatcher_exprs = dispatcher_sig.exprs()
+
+    keyset = 'ks & c10::after_func_keyset'
+    return_type = dispatcher_sig.returns_type().remove_const_ref().cpp_type()
+
+    api_name = CppSignatureGroup.from_native_function(
+        f, method=False, fallback_binding=f.manual_cpp_binding).most_faithful_signature().name()
+
+    # view op case
+    assert is_view_op(f)
+
+    if f.tag is not None and f.tag is Tag.inplace_view:
+        # This op is both an inplace op AND a view op.
+        # See Note [Functionalization Pass - Inplace View Ops] for details.
+        # I currently have the view meta call into the out-of-place variant of the view, to avoid
+        # having to define an extra ~20 inplace {view}_inverse_ functions.
+        # Most view ops don't have NativeFunctionGroup's both, because we don't define out= variants for view ops.
+        # For now, I'm dangerously assuming that every inplace-view op has a corresponding out-of-place view op,
+        # with the same name but the trailing underscore removed.
+        assert f.func.kind() is SchemaKind.inplace
+        api_name = api_name[:-1]
+
+
+    non_self_args = dispatcher.jit_arguments(f.func, skip_self=True)
+    for a in non_self_args:
+        # There currently aren't any views that take extra tensor args aside from self.
+        # If we eventually have any, the codegen will need extra logic to support it - for example,
+        # when we call the at::meta:: API, all tensor args need to be converted to the meta device.
+        assert not a.type.is_tensor_like()
+
+    non_self_bindings = [dispatcher.argument(a) for a in non_self_args]
```

Comment:
Just `dispatcher_sig.arguments()`, no need to do it all manually

Reply:
(In particular, excluding self from being in scope is unnecessary, since `non_self_value_ctypes` doesn't request self anyway, so it's harmless if there is a self argument floating around in the context)